core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -263,11 +338,99 @@ private void remove0() {
         }
     }
 
+    private final class CopyOnWriteAttribute<T> implements Attribute<T> {
+
+        private final Attribute<T> attrInRoot;
+
+        @Nullable
+        private Attribute<T> attrInThis;
+
+        CopyOnWriteAttribute(Attribute<T> attrInRoot) {
```

Comment:
I think `root` concept does not make sense a little. 🧐
How about `attrInRoot` -> `original`, `attrInThis` -> `copied`?